### DISCUSSION TITLE
How to handle race condition in Observable.java?

### DISCUSSION THREAD

**Discussion by ilyash-b** [eclipse-edc/Connector#740] [Q&A] [answered]
https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/84bb9597b4ed18ad177fe82e475661a53000d762/spi/src/main/java/org/eclipse/dataspaceconnector/spi/Observable.java#L35-L36

Imagine two threads registering same listener. Both get into `if`, then both register.

Not 100% sure. Can somebody take a look?

**Comment by jimmarino** [comment, marked as answer]
This would likely not happen:

(1) Runtime bootstrap is single-threaded

(2) If a listener was registered after runtime boot, the same object instance would need to be registered from two different threads, which would have to be intentionally done (since it is the same object instance or a poor implementation of .equals()).
